fuse_models/include/fuse_models/common/sensor_proc.h

```diff
@@ -706,6 +706,194 @@ inline bool processDifferentialPoseWithCovariance(
+  const bool validate,
+  fuse_core::Transaction& transaction)
+{
+  if (position_indices.empty() && orientation_indices.empty())
+  {
+    return false;
+  }
+
+  // Convert the poses into tf2_2d transforms
+  tf2_2d::Transform pose1_2d;
+  tf2::fromMsg(pose1.pose.pose, pose1_2d);
+
+  tf2_2d::Transform pose2_2d;
+  tf2::fromMsg(pose2.pose.pose, pose2_2d);
+
+  // Create the pose variables
+  auto position1 = fuse_variables::Position2DStamped::make_shared(pose1.header.stamp, device_id);
+  auto orientation1 =
+    fuse_variables::Orientation2DStamped::make_shared(pose1.header.stamp, device_id);
+  position1->x() = pose1_2d.x();
+  position1->y() = pose1_2d.y();
+  orientation1->yaw() = pose1_2d.yaw();
+
+  auto position2 = fuse_variables::Position2DStamped::make_shared(pose2.header.stamp, device_id);
+  auto orientation2 = fuse_variables::Orientation2DStamped::make_shared(pose2.header.stamp, device_id);
+  position2->x() = pose2_2d.x();
+  position2->y() = pose2_2d.y();
+  orientation2->yaw() = pose2_2d.yaw();
+
+  // Create the delta for the constraint
+  const double sy = ::sin(-pose1_2d.yaw());
+  const double cy = ::cos(-pose1_2d.yaw());
+  double x_diff = pose2_2d.x() - pose1_2d.x();
+  double y_diff = pose2_2d.y() - pose1_2d.y();
+  fuse_core::Vector3d pose_relative_mean;
+  pose_relative_mean <<
+    cy * x_diff - sy * y_diff,
+    sy * x_diff + cy * y_diff,
+    (pose2_2d.rotation() - pose1_2d.rotation()).getAngle();
+
+  // Create the covariance components for the constraint
+  fuse_core::Matrix3d cov;
+  cov <<
+    twist.twist.covariance[0],
+    twist.twist.covariance[1],
+    twist.twist.covariance[5],
+    twist.twist.covariance[6],
+    twist.twist.covariance[7],
+    twist.twist.covariance[11],
+    twist.twist.covariance[30],
+    twist.twist.covariance[31],
+    twist.twist.covariance[35];
+
+  fuse_core::Matrix3d pose_relative_covariance;
+  // For dependent pose measurements p1 and p2, we assume they're computed as:
+  //
+  // p2 = p1 * p12    [1]
+  //
+  // where p12 is the relative pose between p1 and p2, which is computed here as:
+  //
+  // p12 = p1^-1 * p2
+  //
+  // Note that the twist t12 is computed as:
+  //
+  // t12 = p12 / dt
+  //
+  // where dt = t2 - t1, for t1 and t2 being the p1 and p2 timestamps, respectively.
+  //
+  // Therefore, the relative pose p12 is computed as follows given the twist t12:
+  //
+  // p12 = t12 * dt
+  //
+  // The covariance propagation of this equation is:
+  //
+  // C12 = J_t12 * T12 * J_t12^T    [2]
+  //
+  // where T12 is the twist covariance and J_t12 is the jacobian of the equation wrt to t12.
+  //
+  // The jacobian wrt t12 is:
+  //
+  // J_t12 = dt * Id
+  //
+  // where Id is a 3x3 Identity matrix.
+  //
+  // In some cases the twist covariance T12 is very small and it could yield to an ill-conditioned C12 covariance. For
+  // that reason a minimum covariance is added to [2].
+  const auto dt = (pose2.header.stamp - pose1.header.stamp).toSec();
+
+  if (dt < 1e-6)
```

Comment:
Not sure if I should make `1e-6` an argument + ROS param, or if I should use a smaller value. :thinking: